rr_common/src/planner/distance_map.cpp

```diff
@@ -0,0 +1,103 @@
+#include <rr_common/planning/distance_map.h>
+
+namespace rr {
+
+DistanceMap::DistanceMap(ros::NodeHandle nh)
+      : distance_cost_map(), hit_box(ros::NodeHandle(nh, "hitbox")), listener(new tf::TransformListener) {
+    std::string map_topic;
+    assertions::getParam(nh, "map_topic", map_topic);
+    assertions::getParam(nh, "publish_distance_map", publish_distance_map);
+    assertions::getParam(nh, "publish_inscribed_circle", publish_inscribed_circle);
+
+    assertions::getParam(nh, "cost_scaling_factor", cost_scaling_factor, { assertions::greater_eq(0.0) });
+    assertions::getParam(nh, "wall_inflation", wall_inflation, { assertions::greater_eq(0.0) });
+
+    map_sub = nh.subscribe(map_topic, 1, &DistanceMap::SetMapMessage, this);
+    distance_map_pub = nh.advertise<nav_msgs::OccupancyGrid>("distance_map", 1);
+    inscribed_circle_pub = nh.advertise<geometry_msgs::PolygonStamped>("inscribed_circle", 1);
+
+    std::tie(inscribed_circle_radius, inscribed_circle_origin) = hit_box.getForwardInscribedCircle();
+}
+
+double DistanceMap::DistanceCost(const rr::Pose& pose) {
+    auto [mx, my] = this->PoseToGridPosition(pose);
+
+    if (my < 0 || mapMetaData.height <= my || mx < 0 || mapMetaData.width <= mx)
+        return 0.0;
+
+    return distance_cost_map.at<float>(my, mx);
+}
+
+std::pair<unsigned int, unsigned int> DistanceMap::PoseToGridPosition(const rr::Pose& pose) {
+    tf::Pose w_pose = transform * tf::Pose(tf::createQuaternionFromYaw(0),
+                                           tf::Vector3(pose.x + inscribed_circle_origin, pose.y, 0));
+
+    unsigned int mx = std::floor((w_pose.getOrigin().x() - mapMetaData.origin.position.x) / mapMetaData.resolution);
+    unsigned int my = std::floor((w_pose.getOrigin().y() - mapMetaData.origin.position.y) / mapMetaData.resolution);
+
+    return std::make_pair(mx, my);
+}
+
+void DistanceMap::SetMapMessage(const boost::shared_ptr<nav_msgs::OccupancyGrid const>& map_msg) {
+    if (!accepting_updates_) {
+        return;
+    }
+
+    try {
+        listener->waitForTransform(map_msg->header.frame_id, "base_footprint", ros::Time(0), ros::Duration(.05));
+        listener->lookupTransform(map_msg->header.frame_id, "base_footprint", ros::Time(0), transform);
+    } catch (tf::TransformException& ex) {
+        ROS_ERROR_STREAM(ex.what());
+    }
+
+    mapMetaData = map_msg->info;
+
+    // Turn occupancy grid to distance map in meters
+    cv::Mat distance_map(mapMetaData.width, mapMetaData.height, CV_8UC1);
+    memcpy(distance_map.data, map_msg->data.data(), map_msg->data.size() * sizeof(uint8_t));
```

Comment:
`memcpy` smells like a potential bug, but if the memory layouts of the message and the Mat are the same, then I guess it's fine

Reply:
Both are continuous so it works fine, but I can change it to the nested for-loop way if you think `memcpy` will produce problems.